### PR TITLE
Fix pyomo help -s

### LINKED ISSUE #266: GAMSShell broke help -s
`pyomo help -s` lists the serial solver interfaces, then throws:
```
Traceback (most recent call last):
  File "/usr/bin/pyomo", line 11, in <module>
    sys.exit(main())
  File "/usr/lib64/python3.4/site-packages/pyomo/scripting/pyomo_main.py", line 82, in main
    retval = _options.func(_options)
  File "/usr/lib64/python3.4/site-packages/pyomo/scripting/driver_help.py", line 457, in help_exec
    help_solvers()
  File "/usr/lib64/python3.4/site-packages/pyomo/scripting/driver_help.py", line 347, in help_solvers
    if s == 'py' or opt._metasolver:
AttributeError: 'GAMSShell' object has no attribute '_metasolver'
```

### PR DESCRIPTION
**Pull request by gseastream**
## Summary/Motivation:
See issue. Simple fix so that the `pyomo help -s` command runs properly again.

Note that the GAMS solver breaks when you try to use the `pyomo solve model.py --solver=gams` command, because it doesn't have a `config_block` method, which it would if it inherited from `OptSolver`. I got confused quickly when I tried to understand how those worked for the solvers.

## Changes proposed in this PR:
- Add `_metasolver=False` attribute to GAMS and GDPopt solvers
- Fix `logging` vs `logger` typo causing a bug
- Make some of the doc strings longer

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by gseastream** [comment]
Another thing I noted that was mentioned in the output of `pyomo help -s` here:

```
Pyomo also supports solver interfaces that are wrappers around third-
party solver interfaces. These interfaces require a subsolver
specification that indicates the solver being executed.  For example,
the following indicates that the ipopt solver will be used:

   asl:ipopt

The asl interface provides a generic wrapper for all solvers that use
the AMPL Solver Library.

Note that subsolvers can not be enumerated automatically for these
interfaces.  However, if a solver is specified that is not found,
Pyomo assumes that the asl solver interface is being used.  Thus the
following solver name will launch ipopt if the 'ipopt' executable is
on the user's path:

   ipopt
```

(Although now I'm not so sure about what it says about `ipopt` since that interface has apparently since changed)

So I tried it out. Turns out with any solver you can do `opt = SolverFactory("solver:subsolver")`, or equivalently `SolverFactory("solver", solver="subsolver")`. This sets the `opt.options` to contain `solver="subsolver"`. However, if this functionality was not meant for said solver, it just seems to ignore that you did that, for example whether or not I did `SolverFactory("ipopt")` or `SolverFactory("ipopt:whatever")` didn't change anything. I bring this up because the GAMS solver doesn't even let you try this, since the creation routine tries to call `opt.set_options('solver='+_name)` and the GAMS solver doesn't have a `set_options` attribute (again because it doesn't inherit from `OptSolver`). I figured I didn't need to bother implementing anything to allow that since it's not a supported feature anyway plus the GAMS solver doesn't even use it's `options` attribute, but you do get an error/warning if you try to use this subsolver thing.

For anyone interested, here's the whole output of `pyomo help -s` since no one's seen it in almost a year:

<details>
<summary>Click here to expand</summary>

```

Pyomo Solvers and Solver Managers
---------------------------------
Pyomo uses 'solver managers' to execute 'solvers' that perform
optimization and other forms of model analysis.  A solver directly
executes an optimizer, typically using an executable found on the
user's PATH environment.  Solver managers support a flexible mechanism
for asyncronously executing solvers either locally or remotely.  The
following solver managers are available in Pyomo:

    neos       Asynchronously execute solvers on the NEOS server
    phpyro     Specialized PH solver manager that uses pyro
    pyro       Execute solvers remotely using pyro
    serial     Synchronously execute solvers locally

If no solver manager is specified, Pyomo uses the serial solver
manager to execute solvers locally.  The pyro and phpyro solver
managers require the installation and configuration of the pyro
software.  The neos solver manager is used to execute solvers on the
NEOS optimization server.


Serial Solver Interfaces
------------------------
The serial, pyro and phpyro solver managers support the following
solver interfaces:

    asl                  + Interface for solvers using the AMPL Solver
                           Library
    baron                  The BARON MINLP solver
    bilevel_blp_global   + Global solver for continuous bilevel linear
                           problems
    bilevel_blp_local    + Local solver for continuous bilevel linear
                           problems
    bilevel_ld           + Solver for bilevel problems using linear
                           duality
    cbc                    The CBC LP/MIP solver
    conopt                 The CONOPT NLP solver
    cplex                  The CPLEX LP/MIP solver
    cplex_direct           Direct python interface to CPLEX
    cplex_persistent       Persistent python interface to CPLEX
    gams                 * The GAMS modeling language
    gdpopt               * The GDPopt decomposition-based GDP solver
    glpk                   The GLPK LP/MIP solver
    gurobi                 The GUROBI LP/MIP solver
    gurobi_direct          Direct python interface to Gurobi
    gurobi_persistent      Persistent python interface to Gurobi
    ipopt                * The Ipopt NLP solver
    mpec_minlp           + MPEC solver transforms to a MINLP
    mpec_nlp             + MPEC solver that optimizes a nonlinear
                           transformation
    path                   Nonlinear MCP solver
    pico                   The PICO LP/MIP solver
    ps                   * Pyomo's simple pattern search optimizer
    py                   + Direct python solver interfaces
    scip                   The SCIP LP/MIP solver
    xpress                 The XPRESS LP/MIP solver

An asterisk indicates solvers that are currently available to be run
from Pyomo with the serial solver manager. A plus indicates meta-
solvers, that are always available.

Pyomo also supports solver interfaces that are wrappers around third-
party solver interfaces. These interfaces require a subsolver
specification that indicates the solver being executed.  For example,
the following indicates that the ipopt solver will be used:

   asl:ipopt

The asl interface provides a generic wrapper for all solvers that use
the AMPL Solver Library.

Note that subsolvers can not be enumerated automatically for these
interfaces.  However, if a solver is specified that is not found,
Pyomo assumes that the asl solver interface is being used.  Thus the
following solver name will launch ipopt if the 'ipopt' executable is
on the user's path:

   ipopt


NEOS Solver Interfaces
----------------------
The neos solver manager supports solver interfaces that can be
executed remotely on the NEOS optimization server.  The following
solver interfaces are available with your current system
configuration:

    bonmin       Heuristic MINLP solver
    cbc          MILP solver
    conopt       Feasible path NLP solver
    couenne      Exact MINLP solver
    cplex        MILP solver
    filmint      Heuristic MINLP solver
    filter       SQP NLP solver
    ipopt        Interior point NLP solver
    knitro       Convex MINLP solver
    l-bfgs-b     Bound-constrained NLP solver
    lancelot     Augmented Lagrangian NLP solver
    loqo         Interior point NLP solver
    minlp        Heuristic MINLP solver
    minos        SLC NLP solver
    minto        MILP solver
    mosek        Interior point NLP solver
    ooqp         Convex QP solver
    path         Nonlinear MCP solver
    snopt        SQP NLP solver

```

</details>

File: pyomo/solvers/plugins/solvers/GAMS.py [hunk 1]
```diff
@@ -74,11 +74,13 @@ class GAMSDirect(pyomo.common.plugin.Plugin):
...
     """
     pyomo.common.plugin.implements(IOptSolver)
-    pyomo.common.plugin.alias('_gams_direct', doc='The GAMS modeling language')
+    pyomo.common.plugin.alias('_gams_direct',
+        doc='Direct python interface to the GAMS modeling language')
 
     def __init__(self, **kwds):
         self._version = None
         self._default_variable_value = None
+        self._metasolver = False
```
Comment (qtothec) [inline pyomo/solvers/plugins/solvers/GAMS.py]: I thought the GAMS interface is a metasolver?

Comment (gseastream) [reply pyomo/solvers/plugins/solvers/GAMS.py]: To be honest I had no idea so I took a guess. What's a metasolver?

Comment (qtothec) [reply pyomo/solvers/plugins/solvers/GAMS.py]: A solver that functions mostly to call other solvers. Included are the solver interfaces, like `asl` and `py`.

Actually, looking at the list above, I think GDPopt qualifies as a metasolver as well.

Comment (gseastream) [reply pyomo/solvers/plugins/solvers/GAMS.py]: Currently the only `_metasolver=True` solvers are bilevel, mpec, and asl.

Comment (gseastream) [reply pyomo/solvers/plugins/solvers/GAMS.py]: Actually the help message below the solver list says `A plus indicates meta-solvers, that are always available`, but gams is definitely not always available. In the code for this help function, it decides to show a `+` if it's a metasolver before it decides to show a `*` if it's `available`. Currently also your GDPopt returns True for available no matter what.

**Comment by gseastream** [comment]
Ok so funny story, this fix currently either works or breaks in one of two ways depending on python version. `pyomo help -s` works fine on Python 3.4 and 3.5. On Python 2.7, it produces:

```
...

Serial Solver Interfaces
------------------------
The serial, pyro and phpyro solver managers support the following
solver interfaces:

asl + Interface for solvers using the AMPL Solver
Library
baron * The BARON MINLP solver
bilevel_blp_global + Global solver for continuous bilevel linear
problems
bilevel_blp_local + Local solver for continuous bilevel linear
problems
bilevel_ld + Solver for bilevel problems using linear
duality
cbc * The CBC LP/MIP solver
conopt * The CONOPT NLP solver
cplex * The CPLEX LP/MIP solver
cplex_direct Direct python interface to CPLEX
cplex_persistent Persistent python interface to CPLEX
gams * The GAMS modeling language
gdpopt * The GDPopt decomposition-based GDP solver
glpk The GLPK LP/MIP solver
gurobi * The GUROBI LP/MIP solver
gurobi_direct Direct python interface to Gurobi
gurobi_persistent Persistent python interface to Gurobi
ipopt * The Ipopt NLP solver
mpec_minlp + MPEC solver transforms to a MINLP
mpec_nlp + MPEC solver that optimizes a nonlinear
transformation
path Nonlinear MCP solver
pico The PICO LP/MIP solver
ps * Pyomo's simple pattern search optimizer
py + Direct python solver interfaces
scip * The SCIP LP/MIP solver
Traceback (most recent call last):
File "/home/gseastre/pyomo/bin/pyomo", line 11, in <module>
load_entry_point('Pyomo', 'console_scripts', 'pyomo')()
File "/home/gseastre/pyomo/src/pyomo/pyomo/scripting/pyomo_main.py", line 82, in main
retval = _options.func(_options)
File "/home/gseastre/pyomo/src/pyomo/pyomo/scripting/driver_help.py", line 452, in help_exec
help_solvers()
File "/home/gseastre/pyomo/src/pyomo/pyomo/scripting/driver_help.py", line 342, in help_solvers
if s == 'py' or opt._metasolver:
AttributeError: 'TrustRegionSolver' object has no attribute '_metasolver'
```

And on Python 3.6 it produces:

```
...

Note that subsolvers can not be enumerated automatically for these
interfaces.  However, if a solver is specified that is not found,
Pyomo assumes that the asl solver interface is being used.  Thus the
following solver name will launch ipopt if the 'ipopt' executable is
on the user's path:

   ipopt

Traceback (most recent call last):
  File "/ascldap/users/gseastr/Pyomo/python3.6/bin/pyomo", line 11, in <module>
    load_entry_point('Pyomo', 'console_scripts', 'pyomo')()
  File "/home/gseastr/Pyomo/pyomo/pyomo/scripting/pyomo_main.py", line 82, in main
    retval = _options.func(_options)
  File "/home/gseastr/Pyomo/pyomo/pyomo/scripting/driver_help.py", line 452, in help_exec
    help_solvers()
  File "/home/gseastr/Pyomo/pyomo/pyomo/scripting/driver_help.py", line 369, in help_solvers
    kestrel = pyomo.neos.kestrel.kestrelAMPL()
  File "/home/gseastr/Pyomo/pyomo/pyomo/neos/kestrel.py", line 66, in __init__
    self.setup_connection()
  File "/home/gseastr/Pyomo/pyomo/pyomo/neos/kestrel.py", line 87, in setup_connection
    result = self.neos.ping()
  File "/ascldap/users/gseastr/miniconda3/lib/python3.6/xmlrpc/client.py", line 1112, in __call__
    return self.__send(self.__name, args)
  File "/ascldap/users/gseastr/miniconda3/lib/python3.6/xmlrpc/client.py", line 1452, in __request
    verbose=self.__verbose
  File "/ascldap/users/gseastr/miniconda3/lib/python3.6/xmlrpc/client.py", line 1154, in request
    return self.single_request(host, handler, request_body, verbose)
  File "/ascldap/users/gseastr/miniconda3/lib/python3.6/xmlrpc/client.py", line 1166, in single_request
    http_conn = self.send_request(host, handler, request_body, verbose)
TypeError: send_request() takes 4 positional arguments but 5 were given
```

**Comment by gseastream** [comment]
After reinstalling python and testing on multiple machines, I am now only getting the latter error (`TypeError: send_request() ...`) on every version of python on my RHEL linux machine. Windows works fine and a different linux server works fine for all versions. Not sure why there's a discrepency or if we need to do anything about it.

**Comment by gseastream** [comment]
Ok, the issue is in the `pyomo.neos.kestrel` module with handling proxy connections. We implement a custom [ProxiedTransport](https://github.com/Pyomo/pyomo/blob/bb84f26c051f20caf639839bb1cec0188276335a/pyomo/neos/kestrel.py#L50) class that is used when proxy related environment variables are set, which it seems was copied from an online source. However, this class no longer functions properly. The first issue is a discrepancy in number of arguments, but there are deeper issues in the class that are not compatible with the rest of the `xmlrpclib` module. At this point, I'm just here to report the bug, but this explains the different behavior on different machines (it works fine with regular connections).

As for this PR, two questions still remain: are the GAMS and/or GDPopt solvers metasolvers? Furthermore, should we leave in their new definitions of `_metasolver` or rely on the new defensive strategy that allows them to not specify it?

I suppose an additional question is do we want to add a test for the `pyomo help -s` command, since it seems to call a number of things that were secretly broken so it may be helpful to catch those errors. The one caveat for now is that proxy connections don't work so it actually doesn't function correctly (on certain machines) at the moment.

**Pull request merged**